### PR TITLE
github: add a trigger to retrigger clang-tidy with comment

### PR DESCRIPTION
**Pull request by tchaikov**
before this change, clang-tidy is triggered by a pull request. but there are chances that user wants to retrigger it. for jenkins jobs, user can rebuild a job manually. but for workflow, only the developers with write permission can retrigger a workflow. this is not convenient to regular contributors.

so, in this change, another trigger is added, so that user can trigger the clang-tidy workflow with "/clang-tidy" command. the syntax is inspired by IRC commands.

---

no need to backport, because it's an improvement of the CI.

### CODE REVIEW

**Review by yaronkaikov — APPROVED** [review]
Nice !!!

**Comment by tchaikov** [comment]
@yaronkaikov i added another commit to run the clang-tidy on the merge commit, could you take another look?

**Comment by yaronkaikov** [comment]
> @yaronkaikov i added another commit to run the clang-tidy on the merge commit, could you take another look?

Do we have a merge commit at this point? because the PR is opened still, no?

**Comment by tchaikov** [comment]
> > @yaronkaikov i added another commit to run the clang-tidy on the merge commit, could you take another look?
> 
> Do we have a merge commit at this point? because the PR is opened still, no?

this is how github works. when a pull request is created, github creates a couple references for it under the namespace of "pull/<pull-request-number>". like

 - `refs/pull/<pr>/head`
 - `refs/pull/<pr>/merge`
 - `refs/pull/<pr>/base`

that's why one can fetch a pull request with


```console
$ git fetch origin pull/19409/head:pr-19409-head # for the the head
$ git fetch origin pull/19409/head:pr-19409-merge # for the merged
```

**Review by yaronkaikov — APPROVED** [review]

**Comment by lersek** [comment]
@tchaikov Kefu, you are awesome :)

The first commit looks good to me. I'm a beginner with github actions, but I've made an effort to verify the patch against the github documentation, and it looks sane.

The second commit makes me question my [own comment](https://github.com/scylladb/scylladb/pull/20822#issuecomment-2376807359) from earlier. I have two concerns:
1. If we run clang-tidy on the merge commit, but everything else on the PR head, then that creates inconsistency.
2. I'm just realizing that running workflows on the (future) merge commit is a double-edged sword. Yes, if the master branch is broken in some sense, upon which my own branch is based, then running workflows on the merge commit will pick up "concurrent fixes" on the master branch, and allow my branch to pass CI. However, the opposite is true as well: if the master branch suffers a regression *after* I submit my PR, then that breakage will show up in the merge commit, and it will prevent my branch from passing CI.

So... at this point, I'm not sure anymore what I should be asking for. Perhaps I was just lazy in https://github.com/scylladb/scylladb/pull/20822#issuecomment-2376807359, and the right way may indeed be that I need to wait for the fix to land, then rebase manually, then force-push. I've sort of lost confidence in my original request :/

I think the first patch should be merged in any case; it's always useful (IMO) to let contributors rerun workflows. (I don't think unlimited resource usage should be a concern: that "attack vector" already exists, as people can submit as many PRs as they want.)

Opinions?

**Comment by tchaikov** [comment]
> @tchaikov Kefu, you are awesome :)
> 
> The first commit looks good to me. I'm a beginner with github actions, but I've made an effort to verify the patch against the github documentation, and it looks sane.
> 
> The second commit makes me question my [own comment](https://github.com/scylladb/scylladb/pull/20822#issuecomment-2376807359) from earlier. I have two concerns:
> 
>     1. If we run clang-tidy on the merge commit, but everything else on the PR head, then that creates inconsistency.

agreed on "inconsistency". but it's a compromise. sometime a bunch of PRs could suffer from an issue like #20821, and if we manage to fix it promptly, in order to rerun the CI previously failing because of the regression, we have a couple options:

- ignore the failing CI
- try to rerun the test:

  - if we build the head of PR:
  
    - ask the contributors to rebase their changes, and repush
    - ask the maintainers to rebase the PRs, if the contributors allows maintainer to edit their branches from which the PRs were created, and repush
  - if we build the merge commit:
  
    - click into the jenkins job webpage and push rebuild. (assuming we are building the merge commit)

> 
>     2. I'm just realizing that running workflows on the (future) merge commit is a double-edged sword. Yes, if the master branch is broken in some sense, upon which my own branch is based, then running workflows on the merge commit will pick up "concurrent fixes" on the master branch, and allow my branch to pass CI. However, the opposite is true as well: if the master branch suffers a regression _after_ I submit my PR, then that breakage will show up in the merge commit, and it will prevent my branch from passing CI.

yes, the ability to merge into a buggy master allows us to move in a faster pace, but it is, IMHO, not sustainable.

> 
> 
> So... at this point, I'm not sure anymore what I should be asking for. Perhaps I was just lazy in [#20822 (comment)](https://github.com/scylladb/scylladb/pull/20822#issuecomment-2376807359), and the right way may indeed be that I need to wait for the fix to land, then rebase manually, then force-push. I've sort of lost confidence in my original request :/
> 
> I think the first patch should be merged in any case; it's always useful (IMO) to let contributors rerun workflows. (I don't think unlimited resource usage should be a concern: that "attack vector" already exists, as people can submit as many PRs as they want.)
> 
> Opinions?

i dropped the 2nd commit. but i am in favor of building the merge commit, not because it's a perfect solution, but because it is a better compromise. i will send it later in a separate PR.

**Comment by tchaikov** [comment]
v2:

- dropped the commit which uses the merge commit.

**Review by lersek — APPROVED** [review]

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Pull request closed without merging**
